Add a ClearRectangle item type.

mstange opened the pull request:
Fixes #1926.

I haven't extensively tested this yet. This draws ClearRectangle items as Rectangle primitives with opaque black and operator destination out.
I'm not really sure how to optimize the case where we don't need blending, because the color is set on the primitive before we decide how to blend it. But if there's no blending, we need to set the color to transparent black instead of opaque black.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/1929)
<!-- Reviewable:end -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1924) made this pull request unmergeable. Please resolve the merge conflicts.

mstange commented:
Somehow this change causes my back button's white fill to ignore the rounded clip. Will investigate.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1937) made this pull request unmergeable. Please resolve the merge conflicts.

kvark commented on webrender/src/border.rs:
```diff
@@ -385,7 +385,7 @@ impl FrameBuilder {
                     clip_and_scroll,
                     &info,
-                    &border.top.color,
+                    &FillOrClear::Fill(border.top.color),
```
"XxxxOrYyy" pattern sounds a bit awkward to me. Perhaps, we could go with something like `Fill::Color` versus `Fill::Clear` ?

kvark commented on webrender/src/frame_builder.rs:
```diff
@@ -616,17 +616,23 @@ impl FrameBuilder {
-            return;
-        }
+        let prim = match operation {
+            &FillOrClear::Fill(ref color) => {
```
Could also be `RectangleContent::Fill/Clear`

mstange replied:
I went with this suggestion, `RectangleContent::Fill/Clear`. Thanks!

kvark commented on webrender/src/frame_builder.rs:
```diff
@@ -616,17 +616,23 @@ impl FrameBuilder {
+            &FillOrClear::Fill(ref color) => {
+                // Don't add transparent rectangles to the draw list, but do consider them for hit
+                // testing. This allows specifying invisible hit testing areas.
+                if color.a == 0.0 {
```
should probably move this condition up into the matching, maybe even turn it into an if/let i.e.
```rust
if let &FillOrClear::Fill(ColorF{a: 0.0, ..}) = operation {
   (something something return)
}
let prim = RectanglePrimitive { operation: *operation };
```

mstange commented:
I've addressed the comments. I added a TODO for the missing optimization but didn't address it. I suppose I could make `AlphaRenderItem::add_to_batch` (which knows the chosen blend mode, and whether blending is needed) write a piece of information into the data that's passed to the shader, and then I could modify the rectangle shader to respect that piece of information to switch between opaque black and transparency... but that seems awful.

r? @glennw

glennw commented on webrender/src/frame.rs:
```diff
@@ -442,13 +443,27 @@ impl<'a> FlattenContext<'a> {
+                }
+            }
+            SpecificDisplayItem::ClearRectangle => {
+                if !self.try_to_add_rectangle_splitting_on_clip(
```
Let's not bother trying to split clear rectangles - we can just directly call ```add_solid_rectangle``` here.

glennw commented on webrender/src/prim_store.rs:
```diff
@@ -138,6 +138,7 @@ pub struct PrimitiveMetadata {
     pub cpu_prim_index: SpecificPrimitiveIndex,
     pub gpu_location: GpuCacheHandle,
     pub clip_task_id: Option<RenderTaskId>,
+    pub is_clear: bool,
```
This should be in the ```RectanglePrimitive``` type since it's not common to all primitives.

mstange replied:
How do I get access to the `RectanglePrimitive` inside `get_blend_mode`?

glennw replied:
Something like:

```let rect = &self.cpu_rectangles[metadata.cpu_prim_index.0];```

should work. The ```TextRun``` primitive in the same function does something similar.

mstange replied:
Thanks!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1954) made this pull request unmergeable. Please resolve the merge conflicts.

mstange commented:
I've addressed the comments.

mstange commented:
r? @glennw

glennw commented:
@bors-servo r+

I think the opacity type for a clear rectangle does work - we can easily change it if required, since this is a new interface.

It would be good to add some wrench reftests as a follow up.

bors-servo commented:
:pushpin: Commit fb8b6b3 has been approved by `glennw`

<!-- @bors-servo r=glennw fb8b6b31014da87bbd42eb9aed51e895885a8030 -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1962) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:lock: Merge conflict

mstange commented:
Rebased. r? @glennw

glennw commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 52da7ba has been approved by `glennw`

<!-- @bors-servo r=glennw 52da7baa62e0f0f67426a6f6ee04332f7bde295c -->

bors-servo commented:
:hourglass: Testing commit 52da7baa62e0f0f67426a6f6ee04332f7bde295c with merge c0194de78ce26106a8497484dc8d159069e3a482...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.6155), [status-travis](https://travis-ci.org/servo/webrender/builds/295168456?utm_source=github_status&utm_medium=notification)
Approved by: glennw
Pushing c0194de78ce26106a8497484dc8d159069e3a482 to master...
